### DISCUSSION TITLE
Roadmap

### DISCUSSION THREAD

**Discussion by NorbertKlockiewicz** [software-mansion/react-native-executorch#2] [Announcements]
We're excited to share our plans for the next chapter of React Native ExecuTorch. This roadmap outlines the key features and improvements we're working toward as we head to [v1.0.0](https://github.com/software-mansion/react-native-executorch/milestone/6). From now we aim to provide smaller but more frequent releases. Features included in [v1.0.0](https://github.com/software-mansion/react-native-executorch/milestone/6) are being developed in parallel and can be released sooner. Alongside code development we are working on exporting additional models which we are publishing on our [huggingface](https://huggingface.co/software-mansion). 

As always, this is a living document—priorities may shift based on community feedback, upstream ExecuTorch developments, and what we learn along the way. This means that additional versions can appear before [v1.0.0](https://github.com/software-mansion/react-native-executorch/milestone/6). We'd love to hear your thoughts in the comments.

---

### Release v0.1.0
Release: November 2024

In this release we offer:
1. **Support for large language models**
   - LLAMA 3.2 1B (QLoRA, Spin-Quant, unquantized)
   - LLAMA 3.2 3B (QLoRA, Spin-Quant, unquantized)
2. **New `useLLM` hook which allows to fetch and load models**

---

### Release v0.2.0
Release: December 2024

In this release:
1. **Support for react native new architecture**
2. **Expanding into Computer Vision**
    - Classification models with `useClassification` hook
    - Object detection models with `useObjectDetection` hook
    - Style transfer models with `useStyleTransfer` hook
3. **_Experimental_**: basic ExecuTorch bindings that allow loading and running ExecuTorch compatible models via `useExecutorchModule` as well as creating tensors from _js_ arrays

---

### Release v0.3.0
Release: March 2025

In this release:
1. **Support the OCR model**
2. **Support for Speech to Text model (Whisper)**
3. **Hookless API**
4. **Supported models benchmarks (in particular RAM usage, memory file size, inference time)**

---

### Release v0.4.0
Release: May 2025

In this release:
1. **Universal LLM runner**
2. **More LLMs**
3. **Embedding models**
4. **Multilingual Whisper**
5. **New languages for OCR**
6. **Image segmentation**
7. **ExecuTorch runtime v0.5.0**
8. **Internal improvements and bug fixes**

---

### Release v0.5.0
Release: Aug 2025

In this release:
1. **JSI rewrite of the code - enabling zero-copy data transfer between native code and JS**
2. **CLIP embedding model**
3. **Significant Whisper performance improvements**
4. **LLM structured output**
5. **Significant text-embeddings performance improvements**

---

### Release v0.6.0
Release: Dec 2025

In this release:
1. **Major performance improvements**
2. **Text-to-Image: stable diffusion support**
3. **Voice Activity Detection**
4. **ExecuTorch 1.0.0**
5. **Android 16kB page support**
6.  **Quantized Whisper models**
7.  **Configurable LLM generation - `temperature` and `topp` parameters**

---

### Release [v0.7.0](https://github.com/software-mansion/react-native-executorch/milestone/3)

This release focuses on expanding language support and introducing vision-language capabilities.

1. **Support for Text-To-Speech**
    - [Kokoro](https://huggingface.co/hexgrad/Kokoro-82M) model support (American English, British English)

2. **OCR improvements**
    - Reducing memory requirements and smaller file sizes thanks to combined weights export.

3. **Tokenizers migration**
   - Migration from huggingface to PyTorch tokenizers to reduce bundle size.

4. **Improved error handling**
    - better, more robust error handling with more detailed error messages
---

### Release [v0.8.0](https://github.com/software-mansion/react-native-executorch/milestone/4)

1. **[Vision camera](https://react-native-vision-camera.com/) support**
    - support for [frame processors](https://react-native-vision-camera.com/docs/guides/frame-processors)

2. **Extended language support for Text-to-Speech**
    - Polish
    - Spanish
    - French
    - Italian
    - Brazilian Portuguese
    - Mandarin Chinese

3. **Liquid Foundation Models support**
    - Export and code support for [LFM2.5 models](https://www.liquid.ai/blog/introducing-lfm2-5-the-next-generation-of-on-device-ai).

4. **Computer Vision models quantization**
    - Export for quantized CV models to support realtime video processing

5. **Extended Computer Vision model support**

8. **VLM support**
---

### Release [v0.9.0](https://github.com/software-mansion/react-native-executorch/milestone/5)

1. **Support for GPU/NPU acceleration**
2. **Speech-to-Speech support with [hibiki model](https://github.com/kyutai-labs/hibiki)**
3. **Segment Anything model support**
4. **Improved transformers based models (mainly LLMs & whisper)**

---

### Release [v1.0.0](https://github.com/software-mansion/react-native-executorch/milestone/6)

1. **Modularization**

    Currently all features are bundled together, leading to large bundle sizes and makes production deployment difficult.
    To solve this we are planning on splitting the library into submodules based on usage scenarios, such as LLM or Computer Vision. This would allow us to reduce the bundle size by excluding OpenCV for LLM bundle and skipping tokenizers for Computer Vision.
2. **Comprehensive testing setup**
3. **Model usage telemetry**

    Currently the only way to monitor the library usage is to check for models downloads on our HuggingFace as well as checking npm stats. We are planning on adding usage reports such as what models are used, generated LLM tokens or inference count. This would allow us to properly monitor community interest and enable us to better adjust for future releases. 

    **We are NOT planning on sharing any kind of personal data, such as images, videos or LLM conversations!**
4. **Web support**
    - utilise WASM and webGPU for inference support in web browsers
5. **Removing Expo dependency**
7. **Improved logging**
9. **Fix crashes resulting from running out of device memory**
10. **Benchmarking framework**

    Develop benchmarking framework for library to allow us for regular benchmarking in case for new devices or different models.
11. **GenAI SDK-compatible provider** 

    Provide a GenAI SDK-compatible provider so users can run on-device ExecuTorch models using the familiar GenAI SDK API they already know.
---

### More plans for the future which aren't assigned to any release yet
- **Utility tools to work with Computer Vision related data**
- **YOLO support**
- **Cloud inference fallback for LLMs** - While edge LLMs are getting smarter by the day, some tasks are better suited to be run on larger, cloud models. For this case we are planning on adding dynamic cloud fallback based on user prompt.

---

## How You Can Help

- **Share feedback** — What features matter most to you? Let us know below.
- **Report issues** — Found a bug? [Open an issue](https://github.com/software-mansion/react-native-executorch/issues).
- **Contribute** — PRs are welcome! Check out our [contributing guide](https://github.com/software-mansion/react-native-executorch/blob/main/CONTRIBUTING.md).
- **Spread the word** — Star the repo, share your projects, write about your experience.

Our mission with React Native ExecuTorch is to empower developers to seamlessly integrate their preferred ML models into React Native applications. Starting with targeted solutions, our strategy is to gradually provide a suite of tools addressing more generalized use cases.

This roadmap is a living document. We encourage the community to engage actively—your contributions and insights are invaluable as we strive to meet and exceed developer needs in the evolving landscape of mobile development.

Feel free to join the discussion, contribute to the projects, and let us know how we can better serve you in future updates. 

Thanks for being part of this journey. On-device AI is moving fast, and we're glad to have you along for the ride. 🚀

*— The React Native ExecuTorch Team*

**Comment by gabimoncha** [comment]
what speech to text models have you considered integrating?

**Reply by NorbertKlockiewicz** [reply]
Hi, we are thinking about Whisper from OpenAI.

**Comment by mergennachin** [comment]
@NorbertKlockiewicz @gabimoncha 

Are you aware of this GitHub issue on HF? https://github.com/huggingface/transformers/issues/32253

We're tracking bunch of popular models to be exported on HF.

For example, you can see a sub-issue for Whisper: https://github.com/huggingface/transformers/issues/32253

**Comment by sskarz** [comment]
How can I utilize the XNNPACK backend on this library?
